src/Microsoft.ML.Featurizers/DateTimeTransformer.cs

```diff
@@ -0,0 +1,834 @@
+                CreateTransformerFromEstimatorNative(estimator, out transformer, out errorHandle);
+
+            private protected override bool DestroyEstimatorHelper(IntPtr estimator, out IntPtr errorHandle) =>
+                DestroyEstimatorNative(estimator, out errorHandle);
+
+            private protected override bool DestroyTransformerHelper(IntPtr transformer, out IntPtr errorHandle) =>
+                DestroyTransformerNative(transformer, out errorHandle);
+
+            [DllImport("Featurizers", EntryPoint = "DateTimeFeaturizer_CreateTransformerSaveData", CallingConvention = CallingConvention.Cdecl), SuppressUnmanagedCodeSecurity]
+            private static extern bool CreateTransformerSaveDataNative(TransformerEstimatorSafeHandle transformer, out IntPtr buffer, out IntPtr bufferSize, out IntPtr error);
+            private protected override bool CreateTransformerSaveDataHelper(out IntPtr buffer, out IntPtr bufferSize, out IntPtr errorHandle) =>
+                CreateTransformerSaveDataNative(_transformerHandler, out buffer, out bufferSize, out errorHandle);
+        }
+
+        #endregion
+
+        private sealed class Mapper : MapperBase
+        {
+
+            #region Class data members
+
+            private readonly DateTimeTransformer _parent;
+            private ConcurrentDictionary<long, TimePoint> _cache;
+            private ConcurrentQueue<long> _oldestKeys;
+
+            #endregion
+
+            public Mapper(DateTimeTransformer parent, DataViewSchema inputSchema) :
+                base(parent.Host.Register(nameof(Mapper)), inputSchema, parent)
+            {
+                _parent = parent;
+                _cache = new ConcurrentDictionary<long, TimePoint>();
+                _oldestKeys = new ConcurrentQueue<long>();
+            }
+
+            protected override DataViewSchema.DetachedColumn[] GetOutputColumnsCore()
+            {
+                var columns = new List<DataViewSchema.DetachedColumn>();
+
+                foreach (DateTimeEstimator.ColumnsProduced column in Enum.GetValues(typeof(DateTimeEstimator.ColumnsProduced)))
+                    if (_parent._columnsToDrop == null || !_parent._columnsToDrop.Contains(column))
+                    {
+                        columns.Add(new DataViewSchema.DetachedColumn(_parent._column.Prefix + column.ToString(),
+                            ColumnTypeExtensions.PrimitiveTypeFromType(column.GetRawColumnType())));
+                    }
+
+                return columns.ToArray();
+            }
+
+            private Delegate MakeGetter<T>(DataViewRow input, int iinfo)
+            {
+                ValueGetter<T> result = (ref T dst) =>
+                {
+                    long dateTime = default;
+                    var getter = input.GetGetter<long>(input.Schema[_parent._column.Source]);
```

Comment:
The `GetGetter` method should be called outside of this delegate and cached. Doing this every time a value is requested isn't great for performance.

Reply:
I moved the `GetGetter` outside the delegate following the pattern that the other transformers follow. Is .net able to optimize it if its done that way?

Reply:
See https://github.com/dotnet/machinelearning/blob/master/docs/code/IDataViewImplementation.md#why-getgetter